### ISSUE TITLE
(About Single Operator) why singleSubscriber.isUnsubscribed() & subscription.isUnsubscribed() are all false?

### ISSUE THREAD

**Issue by jdsjlzx** [ReactiveX/RxJava#4759]
The code is as follows:

```
public static void test1() {

    Single<String> single = Single.create(new OnSubscribe<String>() {

        @Override
        public void call(SingleSubscriber<? super String> subscriber) {
            subscriber.onSuccess("Tom");
        }
    });

    SingleSubscriber<String> singleSubscriber = new SingleSubscriber<String>() {

        @Override
        public void onSuccess(String value) {
            // TODO Auto-generated method stub
            System.out.println("onSuccess value = " + value);
        }

        @Override
        public void onError(Throwable error) {
            System.out.println("onError error : " + error);
        }
    };

    Subscription subscription = single.subscribe(singleSubscriber);

    System.out.println("singleSubscriber.isUnsubscribed() ?  " + singleSubscriber.isUnsubscribed());
    System.out.println("subscription.isUnsubscribed() ?  " + subscription.isUnsubscribed());

}
```

The output is as follows:

onSuccess value = Tom
singleSubscriber.isUnsubscribed() ?  false
subscription.isUnsubscribed() ?  false

“http://reactivex.io/documentation/single.html” said：
A Single will call only one of these methods, and will only call it once. Upon calling either method, the Single terminates and the subscription to it ends.

My question is:

why singleSubscriber.isUnsubscribed() & subscription.isUnsubscribed() are all false?

**Comment by akarnokd** [comment]
See #4715.

TL;DR: `SingleSubscriber` is an abstract class and there is no `SafeSingleSubscriber` that calls `unsubscribe` for you so unless you call it manually from your implementation of `onSuccess` and `onComplete`, `isUnsubscribe` won't return null.

**Comment by jdsjlzx** [comment]
Website said:
A Single will call only one of these methods, and will only call it once. Upon calling either method, the Single terminates and the subscription to it ends.

How do I test him?

Take the liberty to ask，this is a bug?

I carefully watched submitted under this code：
https://github.com/ReactiveX/RxJava/pull/4716/files/bdbd69f6ce7e63d2e43e0fb360d321080b5adf2e

Is Fix the bug?

Thanks.

**Comment by jdsjlzx** [comment]
RxJava version is 1.2.1

**Comment by akarnokd** [comment]
Fix it by:

``` java
    SingleSubscriber<String> singleSubscriber = new SingleSubscriber<String>() {

        @Override
        public void onSuccess(String value) {
            unsubscribe();       // <--------------------------------------------
            System.out.println("onSuccess value = " + value);
        }

        @Override
        public void onError(Throwable error) {
            unsubscribe();       // <--------------------------------------------
            System.out.println("onError error : " + error);
        }
    };
```

**Comment by jdsjlzx** [comment]
RxJava 2.0 will fix this question?

 after onSuccess/onError, Subscription auto to unsubscribe ?

Thank you very much

**Comment by akarnokd** [comment]
2.x has `SingleObserver` which does not implement `Disposable` (2.x equivalent of `Subscription` and thus you can't check its state. Calling `isUnsubscribed()` is the wrong way of checking if a consumer has finished or not anyway.

**Comment by jdsjlzx** [comment]
What is the right way  checking if a consumer has finished ?

**Comment by akarnokd** [comment]
By not checking it but reacting to it. Your `onSuccess` and `onCompleted` will be called eventually. Anything else defeats the purpose of using a reactive library (because you are blocking).

**Comment by jdsjlzx** [comment]
Thank you!

**Comment by RobLewis** [comment]
In Android apps, we're told that if we don't unsubscribe everything before an activity is destroyed, we'll get memory leaks. The documentation of `Single` is unclear on this point: when the subscription "ends" with a call to `onSuccess()` or `onError()`, does that prevent memory leaks, or must we also call `unsubscribe()`?

**Comment by akarnokd** [comment]
In v2, terminal events will release resources, sometimes before (`using`), sometimes after (`observeOn`) the respective onXXX returns. There is no need to call `dispose`. However, you have to loose the reference to the consumer (or its `Disposable`) to be sure (`onTerminateDetach`) in case your consumer captured some external references (the Activity for example). In v1, it's a bit incoherent and I don't remember exactly.

**Comment by neworld** [comment]
Same as with Observable. In most cases, they hold the whole activity for a prolonged time. Sometimes they never complete and hold reference forever. IMHO you should unsubscribe always. It could help you to save other resources as well.
